lldb-eval/runner.cc

```diff
@@ -69,65 +69,78 @@ int FindBreakpointLine(const Runfiles& runfiles,
-  lldb::SBProcess process = target.LaunchSimple(nullptr, nullptr, ".");
+  const char* argv[] = {binary_path.c_str(), nullptr};
+
+  auto bp = target.BreakpointCreateByLocation(
+      source_file.c_str(), FindBreakpointLine(source_path.c_str(), break_line));
+  // Test programs don't perform any I/O, so current directory doesn't
+  // matter.
+  auto process = target.LaunchSimple(argv, nullptr, ".");
 
-  bool running = true;
   lldb::SBEvent event;
-  lldb::SBListener listener = debugger.GetListener();
-
-  while (running) {
-    if (listener.WaitForEvent(kWaitForEventTimeout, event)) {
-      if (!lldb::SBProcess::EventIsProcessEvent(event)) {
-        std::cerr << "Got some random event: "
-                  << lldb::SBEvent::GetCStringFromEvent(event) << std::endl;
-        continue;
-      }
-
-      lldb::StateType state = lldb::SBProcess::GetStateFromEvent(event);
-      if (state == lldb::eStateInvalid) {
-        std::cerr << "process event: "
-                  << lldb::SBEvent::GetCStringFromEvent(event) << std::endl;
-        continue;
-      }
-
-      switch (state) {
-        case lldb::eStateStopped: {
-          auto thread = process.GetSelectedThread();
-          auto stopReason = thread.GetStopReason();
-          if (stopReason == lldb::eStopReasonBreakpoint) {
-            lldb::break_id_t bpId = static_cast<lldb::break_id_t>(
-                thread.GetStopReasonDataAtIndex(0));
-            if (bpId == bp.GetID()) {
-              running = false;
-            } else {
-              std::cerr << "Stopped at unknown breakpoint: " << bpId
-                        << std::endl;
-            }
-          }
-          break;
-        }
-        default:
-          break;
-      }
-
-    } else {
+  auto listener = debugger.GetListener();
+
+  for (;;) {
+    if (!listener.WaitForEvent(kWaitForEventTimeout, event)) {
       std::cerr
           << "Timeout while waiting for the event, kill the process and exit."
           << std::endl;
       process.Destroy();
       exit(1);
     }
-  }
 
-  return process;
+    if (!lldb::SBProcess::EventIsProcessEvent(event)) {
+      std::cerr << "Got some random event: "
+                << lldb::SBEvent::GetCStringFromEvent(event) << std::endl;
+      continue;
+    }
+
+    auto state = lldb::SBProcess::GetStateFromEvent(event);
+    if (state == lldb::eStateInvalid) {
+      std::cerr << "process event: "
+                << lldb::SBEvent::GetCStringFromEvent(event) << std::endl;
+      continue;
+    }
+
+    if (state != lldb::eStateStopped) {
+      continue;
+    }
+
+    auto thread = process.GetSelectedThread();
+    auto stopReason = thread.GetStopReason();
+
+    if (stopReason != lldb::eStopReasonBreakpoint) {
+      continue;
+    }
+
+    auto bpId =
+        static_cast<lldb::break_id_t>(thread.GetStopReasonDataAtIndex(0));
+    if (bpId != bp.GetID()) {
+      std::cerr << "Stopped at unknown breakpoint: " << bpId << std::endl;
```

Comment:
I think here we should also exit with an error. If the process is stopped at a different breakpoint, no one is going to resume it.